File: clients/ocaml/lib/protocol.ml
(* Wire layer for the thin-gate HTTP surface: request bytes, response
   head, chunked bodies, NDJSON framing. Pull-style over a [source]. *)

let max_line_bytes = 16 * 1024 * 1024
let call_path = "/v1/call"
let image_path = "/v1/image"
let transcription_path = "/v1/transcription"
let health_path = "/v1/health"

type error =
  | No_response  (** The peer closed before sending a response head. *)
  | Malformed of string
      (** Bytes that do not form a valid HTTP/1.1 response. *)
  | Line_too_long
      (** One NDJSON line exceeded [max_line_bytes] without a newline. *)

type source = {
  read_line : unit -> string option;
  read : int -> string option;
  close : unit -> unit;
}
(** [read_line] is [None] at EOF; [read n] returns exactly [n] bytes, fewer only
    at EOF, [None] when nothing is left. *)

type head = { status : int; headers : (string * string) list }

let header head name = List.assoc_opt name head.headers

let request ~meth ~path ?body () =
  let base =
    Printf.sprintf "%s %s HTTP/1.1\r\nHost: localhost\r\nConnection: close\r\n"
      meth path
  in
  match body with
  | Some b ->
      Printf.sprintf
        "%sContent-Type: application/json\r\nContent-Length: %d\r\n\r\n%s" base
        (String.length b) b
  | None -> base ^ "\r\n"

let split_once s ch =
  match String.index_opt s ch with
  | None -> None
  | Some i ->
      Some (String.sub s 0 i, String.sub s (i + 1) (String.length s - i - 1))

let parse_status line =
  match String.split_on_char ' ' line with
  | ("HTTP/1.1" | "HTTP/1.0") :: code :: _ -> (
      match int_of_string_opt code with
      | Some status -> Ok status
      | None -> Error (Malformed ("malformed status line: " ^ line)))
  | _ -> Error (Malformed ("malformed status line: " ^ line))

let read_head src =
  match src.read_line () with
  | None -> Error No_response
  | Some status_line -> (
      match parse_status status_line with
      | Error e -> Error e
      | Ok status ->
          let rec headers acc =
            match src.read_line () with
            | None ->
                Error (Malformed "connection closed inside the response head")
            | Some "" -> Ok (List.rev acc)
            | Some line ->
                let acc =
                  match split_once line ':' with
                  | Some (k, v) ->
                      (String.lowercase_ascii (String.trim k), String.trim v)
                      :: acc
                  | None -> acc
                in
                headers acc
          in
          Result.map (fun headers -> { status; headers }) (headers []))

let read_exact src n what =
  match src.read n with
  | Some data when String.length data = n -> Ok data
  | _ -> Error (Malformed ("connection closed inside " ^ what))

let contains_ci haystack needle =
  let h = String.lowercase_ascii haystack in
  let n = String.length needle and l = String.length h in
  let rec go i = i + n <= l && (String.sub h i n = needle || go (i + 1)) in
  go 0

let parse_chunk_size line =
  let hex = match split_once line ';' with Some (h, _) -> h | None -> line in
  match int_of_string_opt ("0x" ^ String.trim hex) with
  | Some n -> Ok n
  | None -> Error (Malformed ("malformed chunk size: " ^ line))

(** Reader of body chunks: each call yields the next decoded piece, [Ok None]
    once the body is complete according to its framing. *)
let body_chunks src head : unit -> (string option, error) result =
  let chunked =
    match header head "transfer-encoding" with
    | Some te -> contains_ci te "chunked"
    | None -> false
  in
  if chunked then
    let finished = ref false in
    fun () ->
      if !finished then Ok None
      else
        match src.read_line () with
        | None -> Error (Malformed "connection closed inside the chunked body")
        | Some size_line -> (
            match parse_chunk_size size_line with
            | Error e -> Error e
            | Ok 0 ->
                let rec trailers () =
                  match src.read_line () with
                  | None | Some "" -> ()
                  | Some _ -> trailers ()
                in
                trailers ();
                finished := true;
                Ok None
            | Ok size -> (
                match read_exact src size "a chunk" with
                | Error e -> Error e
                | Ok data -> (
                    match read_exact src 2 "a chunk terminator" with
                    | Ok "\r\n" -> Ok (Some data)
                    | Ok _ -> Error (Malformed "malformed chunk terminator")
                    | Error e -> Error e)))
  else
    match Option.bind (header head "content-length") int_of_string_opt with
    | Some length ->
        let sent = ref false in
        fun () ->
          if !sent || length = 0 then Ok None
          else (
            sent := true;
            Result.map Option.some (read_exact src length "the body"))
    | None -> (
        let eof = ref false in
        fun () ->
          if !eof then Ok None
          else
            match src.read 8192 with
            | None ->
                eof := true;
                Ok None
            | some -> Ok some)

let read_body src head =
  let next = body_chunks src head in
  let out = Buffer.create 1024 in
  let rec go () =
    match next () with
    | Error e -> Error e
    | Ok None -> Ok (Buffer.contents out)
    | Ok (Some data) ->
        Buffer.add_string out data;
        go ()
  in
  go ()

type framer = { mutable buf : string; cap : int }
(** Splits body bytes into NDJSON documents. ['\n'] is the only terminator;
    blank lines are skipped; the cap bounds an unterminated line in bytes. *)

let framer ?(cap = max_line_bytes) () = { buf = ""; cap }

let feed f bytes =
  let buf = f.buf ^ bytes in
  let rec split buf acc =
    match String.index_opt buf '\n' with
    | None -> (buf, List.rev acc)
    | Some i ->
        let line = String.trim (String.sub buf 0 i) in
        let rest = String.sub buf (i + 1) (String.length buf - i - 1) in
        split rest (if line = "" then acc else line :: acc)
  in
  let rest, lines = split buf [] in
  f.buf <- rest;
  if String.length rest > f.cap then Error Line_too_long else Ok lines

let finish f =
  let tail = String.trim f.buf in
  f.buf <- "";
  if tail = "" then None else Some tail

File: clients/lua/mohdel/protocol.lua
--- Wire layer for the thin-gate HTTP surface: request bytes, response
-- head, chunked bodies, NDJSON framing, event and error shapes.
-- Pure Lua 5.1+. No I/O and no JSON codec of its own: readers are
-- driven through a `source` — `:read_line()` (nil at EOF), `:read(n)`
-- (exactly n bytes; fewer only at EOF; nil when nothing is left) and
-- `:close()` — and decoding goes through an injected `decode` function.
local M = {}

M.MAX_LINE_BYTES = 16 * 1024 * 1024

M.PATHS = {
  call = '/v1/call',
  image = '/v1/image',
  transcription = '/v1/transcription',
  health = '/v1/health',
}

local EVENT_TYPES = { delta = true, idle = true, done = true, error = true }

local function trim(s)
  return (s:gsub('^%s+', ''):gsub('%s+$', ''))
end

function M.typed_error(type_, message, extra)
  local e = { type = type_, message = message, severity = 'error', retryable = false }
  if extra then
    for k, v in pairs(extra) do e[k] = v end
  end
  return e
end

function M.is_error(x)
  return type(x) == 'table' and type(x.type) == 'string' and type(x.message) == 'string'
end

function M.is_event(x)
  return type(x) == 'table' and EVENT_TYPES[x.type] == true
end

function M.request(method, path, body)
  local head = {
    method .. ' ' .. path .. ' HTTP/1.1',
    'Host: localhost',
    'Connection: close',
  }
  if body then
    head[#head + 1] = 'Content-Type: application/json'
    head[#head + 1] = 'Content-Length: ' .. #body
  end
  return table.concat(head, '\r\n') .. '\r\n\r\n' .. (body or '')
end

local function protocol_error(message, detail)
  return M.typed_error('PROTOCOL_HTTP_ERROR', message, detail and { detail = detail } or nil)
end

function M.read_head(src)
  local status_line = src:read_line()
  if status_line == nil then
    error(M.typed_error('NET_ERROR', 'no response from thin-gate', { retryable = true }), 0)
  end
  local status = tonumber(status_line:match('^HTTP/%d%.%d (%d%d%d)'))
  if not status then
    error(protocol_error('malformed status line from thin-gate', status_line), 0)
  end
  local headers = {}
  while true do
    local line = src:read_line()
    if line == nil then
      error(protocol_error('connection closed inside the response head'), 0)
    end
    if line == '' then break end
    local k, v = line:match('^([^:]+):%s*(.-)%s*$')
    if k then headers[k:lower()] = v end
  end
  return status, headers
end

local function read_exact(src, n, what)
  local data = src:read(n)
  if data == nil or #data ~= n then
    error(protocol_error('connection closed inside ' .. what), 0)
  end
  return data
end

--- Iterator over body byte chunks. Ends (returns nil) when the body
-- is complete according to its framing.
function M.body_chunks(src, headers)
  local te = headers['transfer-encoding']
  if te and te:lower():find('chunked', 1, true) then
    local finished = false
    return function()
      if finished then return nil end
      local size_line = src:read_line()
      if size_line == nil then
        error(protocol_error('connection closed inside the chunked body'), 0)
      end
      local size = tonumber(size_line:match('^(%x+)') or '', 16)
      if size == nil then
        error(protocol_error('malformed chunk size', size_line), 0)
      end
      if size == 0 then
        local line
        repeat line = src:read_line() until line == nil or line == ''
        finished = true
        return nil
      end
      local data = read_exact(src, size, 'a chunk')
      if read_exact(src, 2, 'a chunk terminator') ~= '\r\n' then
        error(protocol_error('malformed chunk terminator'), 0)
      end
      return data
    end
  end

  local length = tonumber(headers['content-length'])
  if length then
    local sent = false
    return function()
      if sent or length == 0 then return nil end
      sent = true
      return read_exact(src, length, 'the body')
    end
  end

  local eof = false
  return function()
    if eof then return nil end
    local data = src:read(8192)
    if data == nil then
      eof = true
      return nil
    end
    return data
  end
end

function M.read_body(src, headers)
  local parts = {}
  for chunk in M.body_chunks(src, headers) do
    parts[#parts + 1] = chunk
  end
  return table.concat(parts)
end

--- Splits a byte stream into NDJSON documents. `\n` (0x0A) is the only
-- line terminator; the cap bounds one unterminated line in bytes.
function M.ndjson_framer(decode, max_line_bytes)
  local cap = max_line_bytes or M.MAX_LINE_BYTES
  local buf = ''
  local framer = {}

  function framer.feed(bytes)
    buf = buf .. bytes
    local out = {}
    while true do
      local nl = buf:find('\n', 1, true)
      if not nl then break end
      local line = trim(buf:sub(1, nl - 1))
      buf = buf:sub(nl + 1)
      if line ~= '' then out[#out + 1] = decode(line) end
    end
    if #buf > cap then
      error(M.typed_error('PROTOCOL_INVALID_EVENT',
        'NDJSON line exceeds ' .. cap .. ' bytes without newline'), 0)
    end
    return out
  end

  function framer.finish()
    local tail = trim(buf)
    buf = ''
    if tail ~= '' then return decode(tail) end
    return nil
  end

  return framer
end

function M.error_from_body(body, status, decode)
  local ok, parsed = pcall(decode, body)
  if ok and type(parsed) == 'table' and type(parsed.type) == 'string' then
    return parsed
  end
  return M.typed_error('PROTOCOL_HTTP_ERROR', 'thin-gate returned HTTP ' .. status,
    { retryable = status >= 500 })
end

return M
